util/server/index.ts

```diff
@@ -0,0 +1,149 @@
+  const res = await fetch(`${jsonLocation}/index/all?icon=base64`);
+  const devfilesWithoutSource = (await res.json()) as Devfile[];
+  const devfilesWithSource = devfilesWithoutSource.map((devfile: Devfile) => {
+    devfile.sourceRepo = hostName;
+    return devfile;
+  });
+  return devfilesWithSource;
+};
+
+const getLocalJSON = async (hostName: string, jsonLocation: string): Promise<Devfile[]> => {
+  const devfileRelPath = jsonLocation.split('/');
+  const devfileAbsPath = path.join(process.cwd(), ...devfileRelPath, 'index.json');
+  const devfilesUnparsed = await fs.readFile(devfileAbsPath, 'utf8');
+  const devfilesWithoutSource = JSON.parse(devfilesUnparsed) as Devfile[];
+  const devfilesWithSource = devfilesWithoutSource.map((devfile: Devfile) => {
+    devfile.sourceRepo = hostName;
+    return devfile;
+  });
+  return devfilesWithSource;
+};
+
+const getRemoteYAML = async (devfileName: string, yamlLocation: string): Promise<string> => {
+  const res = await fetch(`${yamlLocation}/devfiles/${devfileName}`, {
+    headers: { 'Accept-Type': 'text/plain' }
+  });
+  const devfileYAML = (await res.text()) as string;
+  return devfileYAML;
+};
+
+const getLocalYAML = async (devfileName: string, yamlLocation: string): Promise<string> => {
+  const devfileYAMLRelPath = `${yamlLocation}/${devfileName}/devfile.yaml`.split('/');
+  const devfileYAMLAbsPath = path.join(process.cwd(), ...devfileYAMLRelPath);
+  const devfileYAML = (await fs.readFile(devfileYAMLAbsPath, 'utf8')) as string;
+  return devfileYAML;
+};
+
+const getENVHosts = () => {
+  const envHosts = process.env.DEVFILE_REGISTRY_HOSTS?.split('|').filter((host) => host !== '');
+
+  let hosts: Host = {};
+
+  if (envHosts?.length) {
+    envHosts.forEach((envHost) => {
+      const [hostName, sourceType, hostLocation] = envHost.split('>');
+
+      if (!(sourceType === 'url' || sourceType === 'stacks')) {
+        throw Error(
+          'The environment variable DEVFILE_REGISTRY_HOSTS can only accept "url" or "stacks"'
+        );
+      }
+
+      hosts = {
+        ...hosts,
+        [hostName]: {
+          [sourceType]: hostLocation
+        }
+      };
+    });
+  }
+
+  return hosts;
+};
+
+export const getDevfilesJSON = async (): Promise<Devfile[]> => {
+  let hosts: Host = await getConfigFileHosts('/config/devfile-registry-hosts.json');
+  hosts = { ...hosts, ...getENVHosts() };
+
+  if (!Object.keys(hosts).length) {
+    hosts = {
+      'devfile.io': {
+        url: 'https://registry.devfile.io'
+      }
+    };
+  }
+
+  let devfiles: Devfile[] = [];
+  await Promise.all(
+    Object.entries(hosts).map(async ([hostName, hostLocation]) => {
+      let extractedDevfiles: Devfile[] = [];
+      if (hostLocation.url) {
+        extractedDevfiles = await getRemoteJSON(hostName, hostLocation.url);
+      }
+      if (hostLocation.stacks) {
+        extractedDevfiles = await getLocalJSON(hostName, hostLocation.stacks);
+      }
```

Comment:
what happens when the json has other stuff like

```json
{
  "Example1": {
    "url": "https://registry.devfile.io"
  },
  "Example2": {
    "dummy": "/stacks"
  }
}
```

we should handle these cases

Reply:
we currently do handle those cases. it just does nothing

Reply:
See how it just goes straight through the if statements? https://github.com/schultzp2020/registry-viewer/blob/561318831b85e91f9e0cb7893373ce2ce00b1379/util/server/index.ts#L100

Reply:
fixed